OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProducerSideServiceImpl.h"
+
+#include <thread>
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_service {
+
+void ProducerSideServiceImpl::OnCaptureStartRequested(CaptureEventBuffer* capture_event_buffer) {
+  CHECK(capture_event_buffer != nullptr);
+  LOG("About to send StartCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = capture_event_buffer;
+  }
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStarted;
+  }
+}
+
+void ProducerSideServiceImpl::OnCaptureStopRequested() {
+  LOG("About to send StopCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStopping;
+
+    // Wait (for a limited amount of time) for all producers to send AllEventsSent or to disconnect.
+    service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                              +[](ServiceState* service_state) {
+                                                return service_state->producers_remaining == 0 ||
+                                                       service_state->exit_requested;
+                                              },
+                                              &service_state_),
+                                          max_wait_for_all_events_sent_);
+    CHECK(service_state_.producers_remaining >= 0);
+    if (service_state_.producers_remaining == 0) {
+      LOG("All CaptureEventProducers have finished sending their CaptureEvents");
+    } else {
+      ERROR(
+          "Stopped receiving CaptureEvents from CaptureEventProducers "
+          "even if not all have sent all their CaptureEvents");
+    }
+    service_state_.capture_status = CaptureStatus::kCaptureFinished;
+    service_state_.producers_remaining = 0;
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
+}
+
+void ProducerSideServiceImpl::OnExitRequest() {
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.exit_requested = true;
+  }
+
+  LOG("Attempting to disconnect from CaptureEventProducers as exit was requested");
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    for (grpc::ServerContext* context : server_contexts_) {
+      // This should cause blocking Reads on ServerReaderWriter to fail immediately.
+      context->TryCancel();
+    }
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
```

Comment:
maybe extract this to ClearEventBuffer?

Reply:
I know this is a minor code duplication, but in this code I really want to avoid extracting code that locks mutexes outside of the already numerous (but necessary) methods that take those mutexes, as doing that distracts from the actual code flow as it requires jumping somewhere else. In other words in this complex synchronized code I want to prioritize following the flow (which is already hard enough) to normal code practices.